include/pybind11/pytypes.h

```diff
@@ -211,9 +211,27 @@ class handle : public detail::object_api<handle> {
 public:
     /// The default constructor creates a handle with a ``nullptr``-valued pointer
     handle() = default;
-    /// Creates a ``handle`` from the given raw Python object pointer
+
+    /// Enable implicit conversion from ``PyObject *`` and ``nullptr``.
+    /// Not using ``handle(PyObject *ptr)`` to avoid implicit conversion from ``0``.
+    template <typename T,
+              detail::enable_if_t<detail::any_of<std::is_same<T, PyObject *>,
+                                                 std::is_same<T, std::nullptr_t>>::value,
+                                  int> = 0>
+    // NOLINTNEXTLINE(google-explicit-constructor)
+    handle(T ptr) : m_ptr(ptr) {}
+
+    /// Enable implicit conversion through ``T::operator PyObject *()``.
+    template <typename T,
+              detail::enable_if_t<
+                  detail::all_of<detail::negation<detail::any_of<std::is_base_of<handle, T>,
+                                                                 std::is_same<T, PyObject *>,
+                                                                 std::is_same<T, PyObject *const>,
+                                                                 std::is_same<T, std::nullptr_t>>>,
```

Comment:
Yeah I was still thinking about the cv cases.
The most important thing is to come up with corresponding unit tests first. Without we (me at least) are certain to get lost in the incomprehensible maze of C++ rules and special cases.

Reply:
"feel" and "probably" comments like this are massively unhelpful.
You gave up really quickly (#4006).
I'm presenting a solution that is already globally tested.
If you think this can be one better, please show how.
With tests.
I'm always happy to accept something that is provably better, and will be happy to globally test that for you.

Reply:
Agreed. Apologies, I didn't intend to come off rude or dismissive of your efforts. I was just making a generic note for future discussions. 

Thanks for providing this really helpful PR so far and testing against all the edge case in PyTorch. I just meant to comment on how surprised I was to find that disabling this troublesome implicit conversion in C++ is so difficult. Thanks for all the effort.

Reply:
@Skylion007 I added all the tests that I could think of (53402563afdaa303c6678f256f17999390800e76) to validate that the only behavior change is to disable the implicit conversion from `0` to `handle`. I left a temporary `#ifdef` in case you have ideas for additional tests, or want to validate yourself that the commented-out code fails to compile before and after this PR.

Between
* the existing unit tests
* the added `operator PyObject *()` test reduced from the wild (pytorch)
* the added `operator PyObject *() const` test
* the added `pure_compile_tests_for_handle_from_PyObject_pointers()` tests
* and global testing

I think we are extremely well covered. Maybe the SFINAE constructs could be polished, but unless we have more additional tests that expose a hole, I'd say this is as good as it gets.

I want to rerun the global testing though, although I don't think the change to `none_of` would break anything.